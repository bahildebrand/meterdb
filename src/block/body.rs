//! Implementation of body keys for block storage.
//!
//! This module implements the key/value format used to store timeseries entries in blocks.
//! It has the following format:
//!
//! +------------+----------+-----------+----------+
//! | Key Len(2) |   Key    | TypeLen(2)| Value    |
//! +------------+----------+-----------+----------+
//! |   2 bytes  | variable |  2 bytes  | variable |
//! +------------+----------+-----------+----------+

use bytes::{BufMut, Bytes, BytesMut};

const MAX_KEY_SIZE: usize = 64;

pub(crate) struct BlockEntry {
    bytes: Bytes,
}

impl BlockEntry {
    const KEY_LEN_SIZE: usize = 2;
    const VAL_TYPE_SIZE: usize = 2;

    pub(crate) fn new<T: Into<BlockValue>>(key_str: &str, val: T) -> Result<Self, ()> {
        let key_len = key_str.as_bytes().len();
        if key_len > MAX_KEY_SIZE {
            return Err(());
        }

        let block_val = val.into();
        let block_size = Self::KEY_LEN_SIZE + Self::VAL_TYPE_SIZE + key_len + block_val.size();
        let type_len = TypeLen::from(&block_val);
        let mut entry_bytes = BytesMut::with_capacity(block_size);

        // cast to u16 is safe here, as we verified max key size above
        entry_bytes.put((block_size as u16).to_be_bytes().as_slice());
        entry_bytes.put(key_str.as_bytes());
        entry_bytes.put(type_len.as_bytes().as_slice());
        entry_bytes.put(block_val.as_bytes().as_slice());

        Ok(Self {
            bytes: entry_bytes.freeze(),
        })
    }

    pub(crate) fn size(&self) -> usize {
        self.bytes.len()
    }

    pub(crate) fn as_bytes(&self) -> &[u8] {
        self.bytes.as_ref()
    }
}

pub enum BlockValue {
    UnsignedInt(u32),
}

impl BlockValue {
    pub(crate) fn size(&self) -> usize {
        match self {
            BlockValue::UnsignedInt(_) => size_of::<u32>(),
        }
    }

    // TODO: Should not be hard coded
    pub(crate) fn as_bytes(&self) -> [u8; 4] {
        match self {
            BlockValue::UnsignedInt(v) => v.to_be_bytes(),
        }
    }
}

impl From<u32> for BlockValue {
    fn from(value: u32) -> Self {
        BlockValue::UnsignedInt(value)
    }
}

struct TypeLen {
    val_type: u8,
    size: u8,
}

impl TypeLen {
    const U32_TYPE: u8 = 0;

    fn as_bytes(&self) -> [u8; 2] {
        [self.val_type, self.size]
    }
}

impl From<&BlockValue> for TypeLen {
    fn from(value: &BlockValue) -> Self {
        let val_type = match value {
            BlockValue::UnsignedInt(_) => Self::U32_TYPE,
        };

        Self {
            val_type,
            size: value.size() as u8,
        }
    }
}

#[cfg(test)]
mod test {
    use std::vec::Vec;

    use super::*;

    use insta::assert_binary_snapshot;
    use rstest::rstest;

    #[test]
    fn test_create_block_entry() {
        let test_key = "test_key";
        let test_val = 42u32;
        let entry = BlockEntry::new(test_key, test_val);
        assert!(entry.is_ok());

        let entry = entry.unwrap();
        let entry_bytes = Vec::from(entry.as_bytes());
        let val_size = BlockValue::from(test_val).size();
        let expected_size = test_key.as_bytes().len()
            + BlockEntry::KEY_LEN_SIZE
            + BlockEntry::VAL_TYPE_SIZE
            + val_size;

        assert_eq!(entry_bytes.len(), expected_size);
        assert_binary_snapshot!(".bin", entry_bytes);
    }

    #[test]
    fn block_entry_large_key() {
        let large_key = "a".repeat(MAX_KEY_SIZE + 1);
        let entry = BlockEntry::new(&large_key, 42u32);
        assert!(entry.is_err());
    }

    #[rstest]
    #[case(BlockValue::UnsignedInt(42), size_of::<u32>())]
    fn test_block_value_size(#[case] value: BlockValue, #[case] expected_size: usize) {
        assert_eq!(value.size(), expected_size);
    }
}
