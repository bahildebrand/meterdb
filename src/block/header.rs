use crc_any::{CRCu8, CRCu32};
use heapless::{CapacityError, Vec};
use thiserror::Error;

const MAGIC: u16 = 0xBEEF;
const VERSION: u8 = 0;
pub(crate) const HEADER_LEN: usize = 16;

const ERASED_FLAGS: u8 = 0xFF;
const IN_USE_FLAGS: u8 = 0xFE;
const DIRTY_FLAGS: u8 = 0xFC;

pub(crate) struct Header {
    magic: u16,
    header_crc: u8,
    version: u8,
    flags: u8,
    _reserved: u8,
    pub seq: u16,
    pub len: u32,
    body_crc: u32,
}

impl Header {
    const HEADER_CRC_OFFSET: usize = 2;

    pub(crate) fn as_bytes(&self) -> Result<Vec<u8, HEADER_LEN>, HeaderError> {
        let mut header_bytes = Vec::new();

        header_bytes.extend_from_slice(self.magic.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.header_crc.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.version.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.flags.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self._reserved.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.seq.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.len.to_be_bytes().as_slice())?;
        header_bytes.extend_from_slice(self.body_crc.to_be_bytes().as_slice())?;

        let mut crc8 = CRCu8::crc8();
        crc8.digest(&header_bytes[(Self::HEADER_CRC_OFFSET + 1)..]);
        header_bytes[Self::HEADER_CRC_OFFSET] = crc8.get_crc();

        Ok(header_bytes)
    }

    pub(crate) fn calc_crc(&mut self, bytes: &[u8]) {
        let mut crc32 = CRCu32::crc32d();
        crc32.digest(bytes);
        self.body_crc = crc32.get_crc();
    }
}

impl Default for Header {
    fn default() -> Self {
        Self {
            magic: MAGIC,
            header_crc: 0,
            version: VERSION,
            flags: IN_USE_FLAGS,
            _reserved: 0,
            seq: 0,
            len: HEADER_LEN as u32,
            body_crc: 0,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) enum Flags {
    Erased,
    InUse,
    Dirty,
}

impl From<Flags> for u8 {
    fn from(flags: Flags) -> Self {
        match flags {
            Flags::Erased => ERASED_FLAGS,
            Flags::InUse => IN_USE_FLAGS,
            Flags::Dirty => DIRTY_FLAGS,
        }
    }
}

impl TryFrom<u8> for Flags {
    type Error = HeaderError;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            ERASED_FLAGS => Ok(Flags::Erased),
            IN_USE_FLAGS => Ok(Flags::InUse),
            DIRTY_FLAGS => Ok(Flags::Dirty),
            _ => Err(HeaderError::InvalidFlags(value)),
        }
    }
}

#[derive(Error, Debug)]
pub enum HeaderError {
    #[error("invalid flags: {}", .0)]
    InvalidFlags(u8),
    #[error("not enough capacity for header")]
    HeaderCapacity(#[from] CapacityError),
}

#[cfg(test)]
mod tests {
    use super::*;

    use rstest::rstest;

    #[rstest]
    #[case(Flags::Erased, ERASED_FLAGS)]
    #[case(Flags::InUse, IN_USE_FLAGS)]
    #[case(Flags::Dirty, DIRTY_FLAGS)]
    fn test_int_to_flags(#[case] flags: Flags, #[case] flags_raw: u8) {
        assert_eq!(Flags::try_from(flags_raw).unwrap(), flags);
    }

    #[test]
    fn test_int_to_flags_failure() {
        for flags_raw in 0..=255 {
            if flags_raw != ERASED_FLAGS && flags_raw != IN_USE_FLAGS && flags_raw != DIRTY_FLAGS {
                assert!(Flags::try_from(flags_raw).is_err());
            }
        }
    }

    #[rstest]
    #[case(ERASED_FLAGS, Flags::Erased)]
    #[case(IN_USE_FLAGS, Flags::InUse)]
    #[case(DIRTY_FLAGS, Flags::Dirty)]
    fn test_flags_to_int(#[case] flags_raw: u8, #[case] flags: Flags) {
        assert_eq!(u8::from(flags), flags_raw);
    }

    #[test]
    fn test_header_as_bytes() {
        let expected = [
            0xBE, 0xEF, // magic
            0x02, // header_crc
            0x00, // version
            0xFE, // flags
            0x00, // _reserved
            0x00, 0x01, // seq
            0x00, 0x00, 0x04, 0x00, // len
            0x00, 0x01, 0xE2, 0x40, // body_crc
        ];

        let header = Header {
            seq: 1,
            len: 1024,
            body_crc: 123456,
            ..Default::default()
        };
        let bytes = header.as_bytes().unwrap();
        assert_eq!(bytes.len(), HEADER_LEN);
        assert_eq!(bytes.as_slice(), &expected);
    }

    #[test]
    fn test_crc32() {
        let mut header = Header::default();
        let data = b"hello world";
        header.calc_crc(data);
        assert_eq!(header.body_crc, 0x56C8F614);
    }
}
