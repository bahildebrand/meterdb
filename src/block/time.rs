use chrono::{DateTime, Utc};

pub trait Timestamp {
    fn now(&self) -> Result<DateTime<Utc>, &'static str>;
}

#[cfg(any(feature = "std", test))]
pub struct StdTimestamp;

#[cfg(any(feature = "std", test))]
impl Timestamp for StdTimestamp {
    fn now(&self) -> Result<DateTime<Utc>, &'static str> {
        Ok(Utc::now())
    }
}
