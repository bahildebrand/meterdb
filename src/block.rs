use crate::block::{
    body::{BlockEntry, BlockValue},
    header::Header,
};

mod body;
mod header;

pub(crate) struct Block<const N: usize> {
    header: Header,
    mem: [u8; N],
}

impl<const N: usize> Block<N> {
    pub(crate) fn new() -> Self {
        let header = Header::default();

        Self {
            header,
            mem: [0xFF; N],
        }
    }

    pub(crate) fn add_reading<T: Into<BlockValue>>(&mut self, key: &str, val: T) -> Result<(), ()> {
        let block_entry = BlockEntry::new(key, val)?;
        let header_len = self.header.len as usize;
        if header_len + block_entry.size() > N {
            // TODO: implement block swapping/writing
            return Err(());
        }

        self.header.len += block_entry.size() as u32;
        self.mem[header_len..(header_len + block_entry.size())]
            .copy_from_slice(block_entry.as_bytes());

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use crate::block::header::HEADER_LEN;

    use super::*;

    #[test]
    fn test_add_reading() {
        let mut block: Block<1024> = Block::new();

        assert_eq!(block.header.len as usize, HEADER_LEN);

        let key = "test";
        let test_val = 42u32;

        block.add_reading(key, test_val).unwrap();

        let block_entry = BlockEntry::new(key, test_val).unwrap();
        let expected_size = block_entry.size() + HEADER_LEN;
        assert_eq!(block.header.len as usize, expected_size);
    }

    #[test]
    fn test_add_reading_overflow() {
        let mut block: Block<HEADER_LEN> = Block::new();

        let key = "test";
        let test_val = 42u32;

        // Try to add one more entry and check for overflow
        let result = block.add_reading(key, test_val);
        assert!(result.is_err());
    }
}
