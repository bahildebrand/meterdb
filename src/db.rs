use thiserror::Error;

use crate::block::{Block, BlockBodyError, BlockEntry, BlockError, BlockValue};

pub struct Db<const N: usize, W: DbWriter> {
    cur_block: Block<N>,
    writer: W,
}

impl<const N: usize, W: DbWriter> Db<N, W> {
    pub fn new(writer: W) -> Db<N, W> {
        let cur_block = Block::new();

        Self { cur_block, writer }
    }

    pub fn add_reading<T: Into<BlockValue>>(&mut self, key: &str, val: T) -> Result<(), DbError> {
        let block_entry = BlockEntry::new(key, val)?;
        let res = self.cur_block.add_reading(&block_entry);
        if let Err(BlockError::BlockFull) = res {
            self.flush()?;
            self.cur_block.add_reading(&block_entry)?;

            return Ok(());
        }

        res.map_err(|_| DbError::KeyWrite)
    }

    pub fn flush(&mut self) -> Result<(), DbError> {
        self.cur_block.write_header()?;
        self.writer.write_block(self.cur_block.as_bytes())?;
        self.cur_block.reset();

        Ok(())
    }
}

pub trait DbWriter {
    fn write_block(&mut self, block: &[u8]) -> Result<(), DbError>;
}

#[derive(Error, Debug)]
pub enum DbError {
    #[error(transparent)]
    Block(#[from] BlockError),
    #[error(transparent)]
    Body(#[from] BlockBodyError),
    #[error("failed to write key")]
    KeyWrite,
    #[error("Failed to write block")]
    BlockWrite,
}

#[cfg(test)]
mod test {
    use insta::assert_binary_snapshot;

    use super::*;

    use std::sync::{Arc, Mutex};
    use std::vec::Vec;
    use std::{println, vec};

    struct TestWriter {
        storage: Arc<Mutex<TestBlockStorage>>,
    }

    impl TestWriter {
        pub fn new(storage: Arc<Mutex<TestBlockStorage>>) -> Self {
            Self { storage }
        }
    }

    impl DbWriter for TestWriter {
        fn write_block(&mut self, block: &[u8]) -> Result<(), DbError> {
            let mut storage = self.storage.lock().unwrap();

            storage.write_block(block)
        }
    }

    struct TestBlockStorage {
        pub blocks: Vec<Vec<u8>>,
        pub block_size: usize,
        pub block_count: usize,
        cur_block: usize,
    }

    impl TestBlockStorage {
        pub fn new(block_size: usize, block_count: usize) -> Self {
            let blocks = (0..block_count)
                .map(|_| vec![0; block_size])
                .collect::<Vec<_>>();

            Self {
                blocks,
                block_size,
                block_count,
                cur_block: 0,
            }
        }

        pub fn dump_blocks(&self) -> Vec<u8> {
            let mut all_blocks = Vec::with_capacity(self.block_size * self.block_count);

            for block in &self.blocks {
                all_blocks.extend_from_slice(block.as_slice());
                println!("{:?}", block);
            }

            all_blocks
        }

        fn write_block(&mut self, block: &[u8]) -> Result<(), DbError> {
            let cur_block = &mut self.blocks[self.cur_block];
            self.cur_block += 1;
            println!("{:?}", block);

            cur_block.copy_from_slice(block);

            Ok(())
        }
    }

    #[test]
    fn test_block_write_and_flush() {
        const BLOCK_SIZE: usize = 64;

        let storage = TestBlockStorage::new(BLOCK_SIZE, 2);
        let storage = Arc::new(Mutex::new(storage));
        let writer = TestWriter::new(storage.clone());
        let mut db: Db<BLOCK_SIZE, _> = Db::new(writer);

        let test_key = "test_key";
        let test_val = 42u32;

        db.add_reading(test_key, test_val).unwrap();
        db.flush().unwrap();

        let all_blocks = storage.lock().unwrap().dump_blocks();
        assert_binary_snapshot!(".bin", all_blocks);
    }

    #[test]
    fn test_block_overflow() {
        const BLOCK_SIZE: usize = 64;

        let storage = TestBlockStorage::new(BLOCK_SIZE, 2);
        let storage = Arc::new(Mutex::new(storage));
        let writer = TestWriter::new(storage.clone());
        let mut db: Db<BLOCK_SIZE, _> = Db::new(writer);

        let test_key = "e".repeat(32);
        let test_val = 42u32;

        // write key twice, this should overflow the block
        db.add_reading(&test_key, test_val).unwrap();
        db.add_reading(&test_key, test_val).unwrap();
        db.flush().unwrap();

        let storage = storage.lock().unwrap();

        assert_eq!(storage.cur_block, 2);

        let all_blocks = storage.dump_blocks();
        assert_binary_snapshot!(".bin", all_blocks);
    }
}
