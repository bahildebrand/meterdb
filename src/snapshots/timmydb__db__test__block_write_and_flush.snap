---
source: src/db.rs
expression: all_blocks
extension: bin
snapshot_kind: binary
---
